src/charm.py

```diff
@@ -74,16 +75,29 @@ def sha256(hashable) -> str:
     return hashlib.sha256(hashable).hexdigest()
 
 
+def pull(container, path, *, default: Optional[str] = None) -> Optional[str]:
+    """Pull file from container (without raising FileNotFoundError).
+
+    Returns:
+        File contents if exists; `default` otherwise.
+    """
+    try:
+        return container.pull(path, encoding="utf8").read()
+    except FileNotFoundError:
+        return default
+
+
+def push(container, path, contents):
```

Comment:
Why was this moved to the top level? Nothing else will use it, it forces readers/authors to jump around the code, consider whether or not this is doing something different than `ops.model.Container.push` (it is not), and couldn't meaningfully be used from elsewhere without constructing a [`Container`](https://github.com/canonical/operator/blob/main/ops/model.py#L1748) object by hand, which expects a `_ModelBackend` and to be able to find either an `ops.pebble.Client` object or the socket anyway.

Reply:
Unlike with `pull`, the arguments here are indeed weaker:

- Whenever we push here, we always use `make_dirs=True`.
- For symmetry with the `pull`.

Reply:
I don't think we should use `pull` either...